Sqlite improvements

pmatilai opened the pull request:
Misc improvements:
- handle DBI_CREATED flag correctly for all tables in sqlite (for automatic index rebuilds)
- let sqlite handle header number allocation (for less overflow-prone algorithm, at least in theory)

pmatilai commented:
FWIW, the DBI_CREATED thing was the last actually missing feature in the sqlite backend, I believe it's now on-par with bdb feature-wise. Which is not to say it's complete - it is not.

mlschroe commented:
Just FYI, recycling previously used header ids will break libsolv's `repo_add_rpmdb()` function, which uses the packages from an old solv file if the header id matches.

pmatilai commented:
Hmm, header ids being stable is not a safe assumption regardless of this, they will get renumbered and recycling can occur on --rebuilddb.

pmatilai commented:
It's not like sqlite *has to* do it this way now or ever, it's just handy. But id recycling does already happen, this would just make it a little more common (sqlite only recycles on wraparound and if the largest rowid is removed)

mlschroe commented:
The rebuilddb case is handled by stat()ing the database. A rebuilt database will have a different inode number.

(You probably already know this, but sqlite will guarantee increasing numbers when you use the AUTOINCREMENT keyword.)

pmatilai commented:
Okay, you track inode to detect rebuilds, that will cover the rebuilddb case but not the id wraparound one which ultimately needs to be handled anyway.

And yeah I know AUTOINCREMENT, but it comes with a cost.

BTW just realized that this change actually doesn't really change anything at all in the sqlite backend: it recycles the topmost ids as it is now, because it just does ""SELECT MAX(hnum)" + 1 instead of keeping a permanent record like the other backends do.

The real question is of course the wraparound and what to do about it - outsourcing to sqlite doesn't help a bit as it'll only wrap around at INT64_MAX which is 31 bits beyond our reach.

mlschroe commented:
I wouldn't worry too much about that wrap around: if you add a new rpm header every tenth of a second, you'll need 13 years for the hdrid to wrap.

pmatilai commented:
Header id wraparound is not the most acute of our problems sure, but it's also a bit like a "here be dragons" sign at the edge of the map :) Certainly not the only one of its kind either.

pmatilai commented:
But okay, fair enough. Changed it to use AUTOINCREMENT as there's no real pros to doing it that way but has known practical cons: libsolv, and perhaps others assuming the traditional behavior. Thanks @mlschroe for pointing that out. Despite sqlite doc warnings on autoincrement, performance doesn't seem affected at all, probably because its far from a busy use by database standards.

Also added a couple of other minor tweaks to go.
